src/Microsoft.Restier.Core/ApiBuilderExtensions.cs

```diff
@@ -186,48 +183,91 @@ public static ApiBuilder ChainPrevious<T>(this ApiBuilder obj, Func<IServiceProv
+        /// make an instance of <typeparamref name="TImplement"/> be used in multiple instantiations of
+        /// <typeparamref name="TService"/>, which leads to unpredictable behaviors.
+        /// </remarks>
         /// <typeparam name="TService">The service type.</typeparam>
         /// <typeparam name="TImplement">The implementation type.</typeparam>
         /// <param name="obj">The <see cref="ApiBuilder"/>.</param>
         /// <returns>Current <see cref="ApiBuilder"/></returns>
         public static ApiBuilder ChainPrevious<TService, TImplement>(this ApiBuilder obj)
             where TService : class
-            where TImplement : TService
+            where TImplement : class, TService
         {
-            Func<IServiceProvider, TService, TService> factory = null;
+            Ensure.NotNull(obj, "obj");
+
+            Func<IServiceProvider, Func<TService>, TService> factory = null;
 
-            return obj.AddContributor<TService>((sp, next) =>
+            obj.Services.TryAddTransient<TImplement>();
+            return obj.AddContributorNoCheck<TService>((sp, next) =>
             {
                 if (factory != null)
                 {
-                    return factory(sp, next());
+                    return factory(sp, next);
                 }
 
-                // TODO: See if in the future ActivatorUtilities could replace these implementation. For now its
-                // methods can't select constructor appropriately, base on avaialble services.
-                var ret = ChainedService.CreateViaReflection(typeof(TService), typeof(TImplement), sp, next);
-                var value = ret.Item1;
-                var creator = ret.Item2;
+                var instance = sp.GetService<TImplement>();
+                if (instance == null)
+                {
+                    return instance;
+                }
+
+                var nextProperty = typeof(TImplement).GetTypeInfo().GetProperties()
+                    .FirstOrDefault(e => e.SetMethod.IsPublic && e.PropertyType == typeof(TService));
+                if (nextProperty == null)
+                {
+                    factory = (serviceProvider, _) => serviceProvider.GetRequiredService<TImplement>();
+                    return instance;
+                }
 
-                // Create a factory proxy for the first call, to save the expression compilation cost to the 2nd
-                // call, in case the service is a singleton or for any reason only resolved once.
-                factory = (serviceProvider, nextService) =>
+                nextProperty.SetValue(instance, next());
+                factory = (serviceProvider, getNext) =>
                 {
-                    factory = ChainedService<TService>.CreateFactory(creator);
-                    creator = null;
-                    return factory(serviceProvider, nextService);
+                    var serviceProviderParam = Expression.Parameter(typeof(IServiceProvider));
```

Comment:
better add a block comment here to explain what the below chunk of code does. such as:
```
// To build a lambda expression like:
// factory = (sp, next) => 
// {
//     var hook = sp.GetRequiredService<TImplement>();
//     hook.next = next();
//     return hook;
// }
```
Please correct me if my perception is not right:-)